python_modules/dagster/dagster/_serdes/serdes.py

```diff
@@ -362,47 +450,70 @@ def __init__(
+        unpacked_dict: Dict[str, UnpackedValue],
         whitelist_map: WhitelistMap,
-        descent_path: str,
+        context: UnpackContext,
     ) -> T_NamedTuple:
         try:
-            storage_dict = self.before_unpack(**storage_dict)
+            unpacked_dict = self.before_construct(context, unpacked_dict)
             unpacked: Dict[str, PackableValue] = {}
-            for key, value in storage_dict.items():
+            for key, value in unpacked_dict.items():
                 loaded_name = self.get_loaded_field_name(field=key)
                 # Naively implements backwards compatibility by filtering arguments that aren't present in
                 # the constructor. If a property is present in the serialized object, but doesn't exist in
                 # the version of the class loaded into memory, that property will be completely ignored.
                 if loaded_name in self.constructor_param_names:
-                    unpack_fn = self.get_field_unpack_fn(field=loaded_name)
-                    unpacked[loaded_name] = unpack_fn(
-                        value, whitelist_map=whitelist_map, descent_path=descent_path
-                    )
+                    # custom unpack regardless of hook vs recursive descent
+                    custom = self.field_serializers.get(loaded_name)
+                    if custom:
+                        unpacked[loaded_name] = custom.process(
+                            value,
+                            whitelist_map=whitelist_map,
+                            context=context,
+                        )
+                    else:
+                        unpacked[loaded_name] = context.assert_no_unknown_values(value)
+
+                # if we are ignoring the field and bottom-up parsing, clear unknown values
+                elif isinstance(context, UnpackContext):
+                    context.clear_ignored_unknown_values(value)
 
             # False positive type error here due to an eccentricity of `NamedTuple`-- calling `NamedTuple`
             # directly acts as a class factory, which is not true for `NamedTuple` subclasses (which act
             # like normal constructors). Because we have no subclass info here, the type checker thinks
             # we are invoking the class factory and complains about arguments.
             return self.klass(**unpacked)  # type: ignore
         except Exception as exc:
-            return self.handle_unpack_error(exc, **storage_dict)
+            value = self.handle_unpack_error(exc, context, unpacked_dict)
+            if isinstance(context, UnpackContext):
+                context.assert_no_unknown_values(value)
+                context.clear_ignored_unknown_values(unpacked_dict)
+            return value
 
-    # Hook: Modify the contents of the loaded dict before it is unpacked into domain objects during
+    # Hook: Modify the contents of the unpacked dict before domain object construction during
```

Comment:
"unpack" has meant object construction up until now-- is there something that distiguishes the "unpacked dict" you're referring to above from the raw deserialized dict?

Reply:
oh nm I guess it's that the children have been constructed already